{% extends "layout.html" %}
{% block body %}
<div class="container">
	<div class="row mt-5">
	  <div class="col-sm">
      <h1>Frequently Asked Questions</h1>
	  </div>
	</div>

    <div class="container" style="font-size: 16px;">
      
        <ul class="media-list">
            <li class="media">
                <div class="media-body">
                    <h4 class="media-heading">What is the definition of "short tandem repeat"?</h4>
<p>
We consider repeats with motif lengths between 1-6bp. E.g. AAAAA, ACACACACAC, CAGCAGCAGCAG, TATCTATCTATCTATCTATC,
 TTTTGTTTTGTTTTGTTTTG, AAAAACAAAAACAAAAAC would all fall under this definition. 
</p>
                    <h4 class="media-heading">What STR tools were used to generate results displayed on WebSTR?</h4>
		    <p>
		      We have helped develop <a href="https://github.com/tfwillems/HipSTR">HipSTR</a> and  <a href="https://github.com/gymreklab/GangSTR">GangSTR</a> for genome-wide genotyping of STRs from 
              next-generation sequencing data. Earlier work (mutation rates and constraint) are based on genotypes 
              obtained using <a href="https://github.com/mgymrek/lobstr-code">lobSTR</a>. More recent work (GTEx eSTRs, imputation results) are based on genotypes obtained using HipSTR and GangSTR.
              Recent dataset called EnsebmleTR incorporates even other STR callers like Illumina's <a href="https://github.com/Illumina/ExpansionHunter">ExpansionHunter</a>. 
            </p>
            <h4 class="media-heading">I can't find an STR locus I am interested in on WebSTR. Why are some STRs are not present in the WebSTR database?</h4>
            <p>
                We are aiming to create the most comprehensive genome-wide dataset of STRs and if the loci you are searching has been previously described in the literature
                it should be available on WebSTR. However, WebSTR is currently split in several subsets (STR panels) that are using different methods to both identify repeats on
                the reference genome and to further genotype repeat variations. Due to the statistical nature of these pipelines and filters used at the different stages of the analysis (for instance 
                excluding overlapping repeats) an important loci could have been filtered out. In this case we aim to manually update important missing loci, so please contact us if you find
                such a case. 
            </p>
            <h4 class="media-heading">How do you handle STRs with multiple/complex motifs such as that found in HTT (CTG-CAG) loci
               or motif variation (such as the many motifs known at RFC1)?
            </h4>
            <p>
                We currently report only one motif for such STRs and it is a consensus motif based on the reference genome. This current limitation in WebSTR
                is due to the differences in how repeat motifs are handled by different genotyping tools. WebSTR reports STRs genotyped by
                many different tools and we for now have taken a decision to simplify this aspect of reporting, but plan to improve this in the future. 
            </p>

          <h4 class="media-heading">Does WebSTR display allele frequency information?</h4>
		    <p>
 Yes, if allele frequency statistics are available they are displayed on the locus level page for several cohorts consisting of European, 
 African, and East Asian ancestry.  
		      </p>
		    <h4 class="media-heading">I have generated information I would like to be displayed in WebSTR. How can I add it to the site?</h4>
		    <p>
		      We would love to include your dataset! Contact mgymrek AT ucsd DOT edu to discuss adding summary level STR statistics or 
          allele frequencies for a different cohort to the site.
		    </p>

        <h4 class="media-heading">I would like to set up my own WebSTR locally or on my University cluster, is it possible?</h4>
		    <p>
		       Yes, we provide all the neccessary data files on request if you would like to do it. Corresponding instructions can be found on Github for
           <a href="https://github.com/gymrek-lab/webstr">frontend</a> and <a href="https://github.com/acg-team/webSTR-API">backend</a> tier of WebSTR.  
		    </p>
                </div>
            </li>
        </ul>
    </div>
{% endblock %}
