{% extends "layout.html" %}
{% block body %}

<!-- Div for the plotly chart -->
<script type="text/javascript">
$(document).ready(function() {
    // Initialize all collapsible sections as expanded
    var coll = document.getElementsByClassName("collapsible");
    for (var i = 0; i < coll.length; i++) {
        coll[i].classList.add("active");
        var content = coll[i].nextElementSibling;
        content.style.display = "block";
        // Attach event listeners to toggle collapse
        coll[i].addEventListener("click", function() {
            this.classList.toggle("active");
            var content = this.nextElementSibling;
            if (content.style.display === "block") {
                content.style.display = "none";
                content.classList.remove("expanded-margin");
            } else {
                content.style.display = "block";
                if (content.classList.contains("tcga-content")) {
                    content.classList.add("expanded-margin");
                }
            }
        });
    }
});
</script>

<div class="container-fluid">
	<div class="row justify-content-center">
	  <div class="col-md-10 col-md-offset-1">
		<br><br>
		<h1>STR chr{{ chrom }}:{{ start }}-{{ end }}</h1>
  
		<button class="collapsible active">Reference sequence</button>
		<div class="content">
		  <p>
			<h4>Repeat unit {{ motif }} repeats {{ copies }} times in the reference sequence.</h4>
			<br>
			{{ strseq|safe }}
		  </p>
		</div>
		{% if gene_name != "" %}
		<button class="collapsible active">Overlapping gene</button>
		<div class="content">
		  <p>{{ gene_name }}, {{ gene_desc }}</p>
		</div>
		{% endif %}
		{% if graphJSONx != {} %}
		<button class="collapsible active">Allele Frequency</button>
		<div class="content">
		  <div id="barchart" class="chart"></div>
		  <script type="text/javascript">
			var graphsx = {{ graphJSONx | safe }};
			var graphslayoutx = {{ graphlayoutx | safe }};
			var barchart = document.getElementById('barchart');
			Plotly.newPlot(barchart, graphsx, graphslayoutx);
		  </script>
		</div>
		{% endif %}
		{% if crc_data != None %}
		<button class="collapsible active">Variation in Colorectal Cancer (TCGA cohort)</button>
		<div class="content tcga-content expanded-margin">
		  <table border="1" cellpadding="5" cellspacing="5" class="tablesorter" style="table-layout: fixed; width: 100%">
			<thead class="tablesorter-header">
			  <th>Parameter</th>
			  <th>Value</th>
			</thead>
			<tbody>
			  <tr>
				<td><b>Total calls</b></td>
				<td>{{ crc_data[0] }}</td>
			  </tr>
			  <tr>
				<td><b>Fraction variable</b></td>
				<td>{{ '%.3f' % crc_data[1]|float }}</td>
			  </tr>
			  <tr>
				<td><b>Average length difference</b></td>
				<td>{{ '%.3f' % crc_data[2]|float }}</td>
			  </tr>
			</tbody>
		  </table>
		</div>
		{% endif %}
		{% if estr != None %}
		<button class="collapsible active">Expression STRs</button>
		<div class="content">
		  <table class="tables1" border="1" cellpadding="4" cellspacing="4" style="table-layout: fixed; width: 100%">
			<colgroup>
			  <col style="width: auto" />
			</colgroup>
			<thead class="tables1">
			  <th>Gene (ENS)</th>
			  <th>Tissue</th>
			  <th>Beta (SE)</th>
			  <th>P-value</th>
			  <th>CAVIAR</th>
			</thead>
			{% for d in estr %}
			<tr>
			  <td>{{ d[1] }} ({{ d[0] }})</td>
			  <td>{{ d[2] }}</td>
			  <td>{{ '%0.2f' % d[3]|float }} ({{ '%0.2f' % d[4]|float }})</td>
			  <td>{{ '%0.2e' % d[5]|float }}</td>
			  <td>{{ '%0.2f' % d[6]|float }}</td>
			</tr>
			{% endfor %}
		  </table>
		  See the <a href="/about">about page</a> for a detailed description of each parameter.
		</div>
		{% endif %}
		<button class="collapsible active">Sequence Data</button>
		<div class="content">
		  {% if seq_data != None %}
		  <table border="1" cellpadding="5" cellspacing="5" class="tablesorter tablesorter-default">
			<thead class="tablesorter-header">
			  <tr>
				<th class="sequence-column">Sequence</th>
				<th>AFR</th>
				<th>AMR</th>
				<th>EAS</th>
				<th>EUR</th>
				<th>SAS</th>
			  </tr>
			</thead>
			<tbody>
			  {% for row in seq_data %}
			  <tr>
				<td class="sequence-column">
				  {% for nucleotide in row['sequence'] %}
				  <span class="color-{% if nucleotide == 'A' %}A{% elif nucleotide == 'T' %}T{% elif nucleotide == 'C' %}C{% elif nucleotide == 'G' %}G{% endif %}">{{ nucleotide }}</span>
				  {% endfor %}
				</td>
				<td>{{ row.get('AFR', '&nbsp;') }}</td>
				<td>{{ row.get('AMR', '&nbsp;') }}</td>
				<td>{{ row.get('EAS', '&nbsp;') }}</td>
				<td>{{ row.get('EUR', '&nbsp;') }}</td>
				<td>{{ row.get('SAS', '&nbsp;') }}</td>
			  </tr>
			  {% endfor %}
			</tbody>
		  </table>
		  {% else %}
		  No sequence data available.
		  {% endif %}
		</div>
		<br>
		{% if mut_data != None %}
		<button class="collapsible active">Mutation and constraint</button>
		<div class="content">
		  <table border="1" cellpadding="5" cellspacing="5" class="tablesorter tablesorter-default">
			<thead class="tablesorter-header">
			  <th>Parameter</th>
			  <th>Value</th>
			</thead>
			<tbody>
			  <tr>
				<td><b>Mutation model: Mutation rate</b></td>
				<td>{{ '%.2e' % mut_data[0]|float }}</td>
			  </tr>
			  <tr>
				<td><b>Mutation model: Beta</b></td>
				<td>{{ '%.3f' % mut_data[1]|float }}</td>
			  </tr>
			  <tr>
				<td><b>Mutation model: P(single step)</b></td>
				<td>{{ '%.3f' % mut_data[2]|float }}</td>
			  </tr>
			  <tr>
				<td><b>Constraint (Z-score)</b></td>
				<td>{{ '%.3f' % mut_data[6]|float }}</td>
			  </tr>
			  <tr>
				<td><b>Stutter model: up</b></td>
				<td>{{ '%.3f' % mut_data[3]|float }}</td>
			  </tr>
			  <tr>
				<td><b>Stutter model: down</b></td>
				<td>{{ '%.3f' % mut_data[4]|float }}</td>
			  </tr>
			  <tr>
				<td><b>Stutter model: p</b></td>
				<td>{{ '%.3f' % mut_data[5]|float }}</td>
			  </tr>
			</tbody>
		  </table>
		  See the <a href="/about">about page</a> for a detailed description of each parameter.
		</div>
		{% endif %}
		{% if imp_data != None %}
		<button class="collapsible active">STR Imputation</button>
		<div class="content">
		  <h4>Locus-level imputation metrics</h4>
		  <table border="1" cellpadding="5" cellspacing="5" class="tableshorter tablesorter-default">
			<thead class="tablesorter-header">
			  <th>Parameter</th>
			  <th>Value</th>
			</thead>
			<tbody>
			  <tr>
				<td><b>SSC - Concordance</b></td>
				<td>{{ '%.2f' % imp_data[0]|float }}</td>
			  </tr>
			  <tr>
				<td><b>SSC - r</b></td>
				<td>{{ '%.2f' % imp_data[1]|float }}</td>
			  </tr>
			  <tr>
				<td><b>1000 (EUR) - Concordance</b></td>
				<td>{{ '%.2f' % imp_data[2]|float }}</td>
			  </tr>
			  <tr>
				<td><b>1000G (EUR) - r</b></td>
				<td>{{ '%.2f' % imp_data[3]|float }}</td>
			  </tr>
			  <tr>
				<td><b>1000 (AFR) - Concordance</b></td>
				<td>{{ '%.2f' % imp_data[4]|float }}</td>
			  </tr>
			  <tr>
				<td><b>1000G (AFR) - r</b></td>
				<td>{{ '%.2f' % imp_data[5]|float }}</td>
			  </tr>
			  <tr>
				<td><b>1000 (EAS) - Concordance</b></td>
				<td>{{ '%.2f' % imp_data[6]|float }}</td>
			  </tr>
			  <tr>
				<td><b>1000G (EAS) - r</b></td>
				<td>{{ '%.2f' % imp_data[7]|float }}</td>
			  </tr>
			</tbody>
		  </table>
		  {% if imp_data != None %}
		  <h4>Allele-level imputation metrics</h4>
		  <table border="1" cellpadding="5" cellspacing="5" class="tableshorter tablesorter-default">
			<thead class="tablesorter-header">
			  <th>Allele (bp diff from hg19)</th>
			  <th>r2</th>
			  <th>P-val</th>
			</thead>
			<tbody>
			  {% for d in imp_allele_data %}
			  <tr>
				<td>{{ d[0] }}</td>
				<td>{{ '%0.2f' % d[1]|float }}</td>
				<td>{{ '%0.3e' % d[2]|float }}</td>
			  </tr>
			  {% endfor %}
			</tbody>
		  </table>
		  {% endif %}
		  See the <a href="/about">about page</a> for a detailed description of each parameter.
		</div>
		{% endif %}
		<br><br><br><br>
	  </div>
	</div>
  </div>

{% endblock %}
