{% extends "layout.html" %}

{% block body %}


<!-- Div for the plotly chart -->
<div id="chart" class="chart"></div>


<script type="text/javascript">
  var chrom = {{chrom | tojson}};
  var graphs = {{graphJSON | safe}};
  var layout = {{layoutJSON | safe}};
  var strids = {{strids | safe}};

  var myPlot = document.getElementById('chart'),
    d3 = Plotly.d3,
    data = graphs,
    layout=layout;

  Plotly.newPlot('chart', data, layout);


  myPlot.on('plotly_click', function(data){
    for(var i=0; i < data.points.length; i++){
        if (data.points[i].curveNumber==1) {
          url="None";
        }
        else {
          url="/locus?STRID="+strids[data.points[i].pointNumber]
        }
    }
    if (url != "None") {
    var win = window.open(url, "_blank");
    win.focus();
    }
});

$(document).ready(function(){
  $('[data-toggle="tooltip"]').tooltip(); 
});

</script>
<style>
mark.red {
    color:#ff0000;
    background: white;}
mark.gray {
    color:gray;
    background: white;}
mark.gold {
    color:gold;
    background: white;}
mark.blue {
    color:blue;
    background: white;}
mark.purple {
    color:purple;
    background: white;}
mark.green {
    color:green;
    background: white;}

</style>

<div id="plotLegend" class="plotfooter"></div>

<!-- Main table of STRs in this region -->

<div class="container-fluid">

  <div class="row justify-content-center">
    <div class="col-md-10 col-md-offset-1">

      {% if genome == 'hg19' %}
      <h3>STR locations found in hg19</h3>  
      <p> <b> Motif length:</b> <mark class="gray"> 1: gray, <mark class="red"> 2: red , <mark class="gold"> 3: gold, <mark class="blue"> 4: blue, <mark class="purple"> 5: purple, <mark class="green"> 6: green </p>
      <h6> Hover over table column headings to display addtional details </h6>  
        <div id="variants_table_container" style>  
      <table border="1" cellpadding="5" cellspacing="5" class="table table-bordered">
        <thead class="tablesorter-header"><th><abbr title="The chromosome, start position, and end position of the STR in the hg19 reference genome"> STR Locus (hg19)</abbr></th><th><abbr title="The motif (repeat unit) of each STR is given in canonical format. The canonical repeat unit is defined as the lexicographically first repeat unit when considering all rotations and strand orientations of the repeat sequence. For example, the canonical repeat unit for the repeat sequence CAGCAGCAGCAG would be AGC."> Motif</th></abbr><th><abbr title="The number of consecutive copies of the repeat motif present in the hg19 reference genome."># copies (hg19)</abbr></th><th><abbr title="eSTRs are STRs whose lengths are linearly associated with expression of nearby genes. Each circle gives the effect size of significant eSTRs identified in the GTEx cohort (Fotsing et al. BiorXiv 2018). Circles are color-coded by tissue. Hover over a circle to see the tissue and target gene of each eSTR. Click on an STR to see more information about eSTR effects.">  eSTRs (FDR<0.1)</abbr>  </th> <th><abbr title="Heterozygosity gives a measure of how polymorphic the STR is. It is computed as 1-(sum of squared frequency of each allele). Values close to 1 indicate highly polymorphic STRs. Values close to 0 indicate an STR that is mostly not variable in the population."> Heterozygosity </abbr></th></thead>
        <tbody>
        {% for d in table %}
        <tr>
          <td><a href="/locus?repeat_id={{d[6]}}" target="_blank"> {{ d[0] }}:{{ d [1] }}-{{ d[2] }}</a></td>
          <td>{{ d[3] }}</td>
          <td>{{'%0.1f' % d[5]|float}}</td>
                <td>{{ d[13] | safe }}</td>
                <td>{{ d[10] | safe }}</td>
          
        </tr>
        {% endfor %}
        </tbody>
      </table>
      {% endif %}

      {% if genome == 'hg38' %}
      <h3>STR locations found in hg38</h3>  
        <div id="variants_table_container" style>  
  
          <table border="1" cellpadding="2" cellspacing="2" class="table table-bordered">
            <thead class="tablesorter-header"><th>
              <abbr title="The chromosome, start position, and end position of the STR in the hg38 reference genome"> STR Locus (hg38)</abbr></th>
              <th><abbr title="The motif (repeat unit) of each STR is given in canonical format. The canonical repeat unit is defined as
                 the lexicographically first repeat unit when considering all rotations and strand orientations of the repeat sequence. For example, the canonical repeat unit for the repeat sequence CAGCAGCAGCAG would be AGC."> Motif</th></abbr>
              <th><abbr title="The number of consecutive copies of the repeat motif present in the hg19 reference genome."># copies (hg38)</abbr></th>
              <th><abbr title="Total calls in TCGA CRC dataset"> Total calls</abbr>  </th> 
              <th><abbr title="Frac variable in TCGA CRC dataset"> Frac. variable </abbr></th>
              <th><abbr title="Avg size difference in TCGA CRC dataset"> Avg size diff </abbr></th></thead>
            </thead>
            <tbody>
            {% for d in table %}
            <tr>
              <td><a href="/locus?repeat_id={{d[0]}}&genome=hg38" target="_blank"> 
                {{ d[1] }}:{{ d [2] }}-{{ d[3] }}</a></td>
              <td>{{ d[5] }}</td>
              <td>{{d[7]}}</td>
              <td>{{ d[12] | safe }}</td>
              <td>{{ d[13] | safe }}</td>
              <td>{{ d[14] | safe }}</td>
            </tr>
            {% endfor %}
            </tbody>
          </table>
          {% endif %}
      </div>
    </div>
  </div>
</div>
{% endblock %}
